### PR TITLE
Issue 112

### PR DESCRIPTION
**Pull request by lkitching**

### CODE REVIEW

**Comment by lkitching** [comment]
@zeginis - This PR is for #112 which lets you specify the language tag for labels to be used for the GraphQL schema (see `:schema-language-label` in config.edn).

**Comment by zeginis** [comment]
Thank you @lkitching!
The PR looks good.

**Comment by zeginis** [comment]
I try to run this PR and get an error at GraphiQL interphase. And autocomplete is not working at GraphiQL

```
Error: dataset_alcohol_related_hospital_discharge_observations_dimensions fields must be an object with field names as keys or a function which returns such an object.
    at invariant (http://localhost:8085/graphiql.js:24809:11)
    at GraphQLInputObjectType._defineFieldMap (http://localhost:8085/graphiql.js:28409:29)
    at GraphQLInputObjectType.getFields (http://localhost:8085/graphiql.js:28400:49)
    at typeMapReducer (http://localhost:8085/graphiql.js:29767:26)
    at Array.reduce (<anonymous>)
    at http://localhost:8085/graphiql.js:29760:36
    at Array.forEach (<anonymous>)
    at typeMapReducer (http://localhost:8085/graphiql.js:29753:27)
    at http://localhost:8085/graphiql.js:29762:20
    at Array.forEach (<anonymous>)
```

**Comment by lkitching** [comment]
@zeginis - I've changed the way the schema is generated, does it work locally for you now?

**Comment by zeginis** [comment]
I get an exception at build time:
```
Exception in thread "main" clojure.lang.ExceptionInfo: Key  not found {}, compiling:(C:\Users\user\AppData\Local\Temp\form-init8512082196078002838.clj:1:107)
	at clojure.lang.Compiler.load(Compiler.java:7526)
	at clojure.lang.Compiler.loadFile(Compiler.java:7452)
	at clojure.main$load_script.invokeStatic(main.clj:278)
	at clojure.main$init_opt.invokeStatic(main.clj:280)
	at clojure.main$init_opt.invoke(main.clj:280)
	at clojure.main$initialize.invokeStatic(main.clj:311)
	at clojure.main$null_opt.invokeStatic(main.clj:345)
	at clojure.main$null_opt.invoke(main.clj:342)
	at clojure.main$main.invokeStatic(main.clj:424)
	at clojure.main$main.doInvoke(main.clj:387)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.main.main(main.java:37)
Caused by: clojure.lang.ExceptionInfo: Key  not found {}
	at clojure.core$ex_info.invokeStatic(core.clj:4739)
	at clojure.core$ex_info.invoke(core.clj:4739)
	at graphql_qb.util$strict_get.invokeStatic(util.clj:115)
	at graphql_qb.util$strict_get.invoke(util.clj:112)
	at graphql_qb.dataset_model$get_dimension_type.invokeStatic(dataset_model.clj:138)
	at graphql_qb.dataset_model$get_dimension_type.invoke(dataset_model.clj:129)
	at graphql_qb.dataset_model$construct_dataset$fn__7378.invoke(dataset_model.clj:147)
	at clojure.core$map$fn__5587.invoke(core.clj:2745)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$concat$fn__5215.invoke(core.clj:717)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$empty_QMARK_.invokeStatic(core.clj:6126)
	at clojure.core$empty_QMARK_.invoke(core.clj:6126)
	at graphql_qb.core$can_generate_schema_QMARK_.invokeStatic(core.clj:20)
	at graphql_qb.core$can_generate_schema_QMARK_.invoke(core.clj:17)
	at clojure.core$filter$fn__5614.invoke(core.clj:2813)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$map$fn__5587.invoke(core.clj:2738)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24)
	at clojure.core.protocols$fn__7835.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__7835.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__7781$G__7776__7794.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6748)
	at clojure.core$into.invokeStatic(core.clj:6815)
	at clojure.core$into.invoke(core.clj:6807)
	at graphql_qb.schema.mapping.labels$get_datasets_enum_mappings.invokeStatic(labels.clj:194)
	at graphql_qb.schema.mapping.labels$get_datasets_enum_mappings.invoke(labels.clj:187)
	at graphql_qb.core$get_datasets_enum_mappings.invokeStatic(core.clj:27)
	at graphql_qb.core$get_datasets_enum_mappings.invoke(core.clj:22)
	at graphql_qb.core$build_schema_context.invokeStatic(core.clj:56)
	at graphql_qb.core$build_schema_context.invoke(core.clj:53)
	at graphql_qb.server$create_server.invokeStatic(server.clj:14)
	at graphql_qb.server$create_server.invoke(server.clj:12)
	at graphql_qb.server$start_server.invokeStatic(server.clj:25)
	at graphql_qb.server$start_server.invoke(server.clj:24)
	at graphql_qb.main$_main.invokeStatic(main.clj:55)
	at graphql_qb.main$_main.doInvoke(main.clj:44)
	at clojure.lang.RestFn.invoke(RestFn.java:482)
	at clojure.lang.Var.invoke(Var.java:403)
	at user$eval157.invokeStatic(form-init8512082196078002838.clj:1)
	at user$eval157.invoke(form-init8512082196078002838.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7062)
	at clojure.lang.Compiler.eval(Compiler.java:7052)
	at clojure.lang.Compiler.eval(Compiler.java:7052)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	... 12 more
```
Attached the data I use. They have been created by Table2Cube
[Table2Cube_data.zip](https://github.com/Swirrl/graphql-qb/files/2277685/Table2Cube_data.zip)

File: resources/get-all-components.sparql [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
+PREFIX dcterms: <http://purl.org/dc/terms/>
+
+SELECT * WHERE {
+  ?ds qb:structure ?dsd .
+  ?dsd a qb:DataStructureDefinition .
+  ?dsd qb:component ?comp .
+  OPTIONAL { ?comp qb:order ?order }
+  OPTIONAL { ?comp qb:codeList ?codelist }
```
Comment (zeginis) [inline resources/get-all-components.sparql]: The qb:codeList can be aither at the ?comp or the ?dimension. We should use the `:codelist-source` from the configuration

File: src/graphql_qb/dataset_model.clj [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      "PREFIX qb: <http://purl.org/linked-data/cube#>"
+      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>"
+      "SELECT * WHERE {"
+      "  ?ds a qb:DataSet ."
+      "  ?ds qb:structure ?dsd ."
+      "  ?dsd a qb:DataStructureDefinition ."
+      "  ?dsd qb:component ?comp ."
+      "  ?comp qb:dimension ?dim ."
+      (string/join "\n" dimension-filters)
+      "  ?comp qb:codeList ?codelist ."
```
Comment (zeginis) [inline src/graphql_qb/dataset_model.clj]: The qb:codeList can be aither at the ?comp or the ?dim. We should use the `:codelist-source` from the configuration

File: src/graphql_qb/dataset_model.clj [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+             {:ds ds-uri
+              :name (util/find-best-language names (config/schema-label-language configuration))}))
+         by-uri)))
+
+(defn find-all-dimensions-query [configuration]
+  (let [dataset-label (config/dataset-label configuration)]
+    (str
+      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>"
+      "PREFIX qb: <http://purl.org/linked-data/cube#>"
+      "SELECT ?dim ?label ?comment WHERE {"
```
Comment (zeginis) [inline src/graphql_qb/dataset_model.clj]: ?comment is not assigned. Do we need the comment?

**Comment by lkitching** [comment]
@zeginis - The error seems to be due to `table2qb` using a different predicate for component codelists. I've created #117 to allow it to be configured.

**Comment by zeginis** [comment]
@lkitching what do you mean with "different predicate for component codelists" ?

**Comment by zeginis** [comment]
I see there is a 
`<http://statistics.gov.scot/data/employment/component/gender> <http://publishmydata.com/def/qb/codesUsed> <http://statistics.gov.scot/data/employment/codes-used/gender> .`
But there is also: 
`<http://statistics.gov.scot/def/dimension/gender> qb:codeList <http://statistics.gov.scot/def/concept-scheme/gender> .`

I tried using at the `config.edn` ` :codelist-source "dimension"` but ut does not work.

**Comment by lkitching** [comment]
@zeginis - I've made the changes you suggested (thanks!), so I'm going to merge this branch unless you have any objections?

**Pull request merged**